### PR TITLE
CentOS Stream plugin

### PR DESCRIPTION
**Pull request by shigaraki0**
@t184256

### CODE REVIEW

File: fingertip/plugins/os/centos-stream.ks [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+rootpw fingertip
+timezone UTC --utc
+
+zerombr
+
+firewall --disabled
+network --hostname={HOSTNAME}
+sshkey --username root '{SSH_PUBKEY}'
+
+%packages --timeout 900
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.ks:18]: please try adding `--excludeWeakdeps` for even less packages

File: fingertip/plugins/os/centos-stream.ks [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+dnf -y config-manager --disable '*-debuginfo' '*-source'
+dnf -y makecache
+dnf -y remove linux-firmware
+dnf -y update
+dnf -y clean packages
+fstrim -av
+%end
+
+poweroff
+
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.ks:68]: empty line at the end of the file

File: fingertip/plugins/os/centos-stream.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import os
+
+import fingertip.machine
+from fingertip.util import path
+
+#[Repos]
+#http://ftp.fi.muni.cz/pub/linux/centos/8-stream/
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:7]: this isn't used here

File: fingertip/plugins/os/centos-stream.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+def main(m=None, fips=False):
+    m = m or fingertip.build('backend.qemu')
+    if hasattr(m, 'qemu'):
+        return m.apply(install_in_qemu, fips=fips)
+    raise NotImplementedError()
+
+
+def install_in_qemu(m=None,
+                    fips=False):
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:20]: Please remove the `fips` parameter, I don't see anyone using it in the close future.
Or allow passing custom stuff to `-append` instead.

File: fingertip/plugins/os/centos-stream.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    """
+    :param bool fips: if True, install system in FIPS mode
+    """
+
+    m = m or fingertip.machine.build('backend.qemu')
+    URL = 'http://mirror.centos.org/centos/8-stream/BaseOS/x86_64/os/'
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:26]: Please move the constant to the top of the file outside of the functions (and maybe the comment will become unnecessary).

File: fingertip/plugins/os/centos-stream.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                                      PROXY=m.http_cache.internal_url)
+        m.expiration.depend_on_a_file(ks)
+        m.http_cache.mock('http://mock/ks', text=ks_text)
+        m.log.info(f'fetching kernel: {URL}/isolinux/vmlinuz')
+        kernel = os.path.join(m.path, 'kernel')
+        m.http_cache.fetch(f'{URL}/isolinux/vmlinuz', kernel)
+        m.log.info(f'fetching initrd: {URL}/isolinux/initrd.img')
+        initrd = os.path.join(m.path, 'initrd')
+        m.http_cache.fetch(f'{URL}/isolinux/initrd.img', initrd)
+        append = ('ks=http://mock/ks console=ttyS0 ' + 'inst.text ' + 'inst.notmux ' + 'inst.cmdline ' +
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:54]: 1) keep lines shorter than 79 characters
2) why do that `'a ' + 'b ' + 'c '` when you can do `'a b c '`?

File: fingertip/plugins/os/centos-stream.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            m.console.sendline(username)
+            m.console.expect('Password: ')
+            m.console.sendline(password)
+            m.console.expect_exact(m.prompt)
+
+        m.login = login
+
+        m.login()
+
+        m.log.info('Centos installation finished')
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:82]: `CentOS`. Or, better, `CentOS Stream`.

File: fingertip/plugins/os/centos-stream.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+        m.hooks.disable_proxy.append(disable_proxy)
+
+        m.hooks.unseal += [lambda: m('systemctl restart NetworkManager'),
+                           lambda: m('nm-online')]
+
+        m.hooks.timesync.append(lambda: m('hwclock -s'))
+
+        m.centos = 8
+        m.release = '8'
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:100]: I don't think this is needed; IIUC, CentOS Stream doesn't have minor releases.

File: fingertip/plugins/os/centos-stream.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        m.dist_git_branch = 'c8s'
+
+        return m
+
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:104]: Empty line at the end of the file.

**Review by t184256 — CHANGES_REQUESTED** [review]
+ rebase on top of current master

File: fingertip/plugins/os/centos-stream.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        m.http_cache.fetch(f'{URL}/isolinux/initrd.img', initrd)
+        append = ('ks=http://mock/ks console=ttyS0 ' +
+                  'inst.text inst.notmux inst.cmdline ' +
+                  f'proxy={m.http_cache.internal_url} ' + f'repo={URL} ')
+        extra_args = ['-kernel', kernel, '-initrd', initrd, '-append', append]
+
+        m.ram.safeguard = '768M'
+        with m.ram('>=4G'):
+            m.qemu.run(load=None, extra_args=extra_args)
+            m.console.expect('Running post-installation scripts')
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.py:49]: Better expect `Installation complete`.

File: fingertip/plugins/os/centos-stream.ks [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+openssh-server
+glibc-minimal-langpack
+glibc-langpack-en
+python3-libselinux
+kexec-tools
+-subscription-manager
+-cronie
+-sg3_*
+-parted
+
```
Comment (t184256) [inline fingertip/plugins/os/centos-stream.ks:28]: remove the empty line, please

**Review by t184256 — CHANGES_REQUESTED** [review]
Seems to work OK. Please address two very minor things and add yourself to CONTRIBUTORS.

**Comment by shigaraki0** [comment]
@t184256

**Comment by t184256** [comment]
LGTM!

**Pull request merged**